File: project/cipherImplementation.cpp
///////////////////////////////////////////////////////////////////////////////////
//                                                                               //
//                                                                               //
//                  written by Irad Nuriel irad9731@gmail.com                    //
//                        written in Jun 4 2021                                  //
//                                                                               //
//                                                                               //
///////////////////////////////////////////////////////////////////////////////////
#include "cipherImplementation.h"
#include "constants.h"
#include <stdint.h>
#include <iostream>
#include <iomanip>
#include <time.h>
using namespace std;


word rotateLeft(word w, int n){
    word mask = 0xFFFFFFFFFFFFFFFF;
    return ((w << n) & mask) | ((w >> (64 - n) & mask));
}



word keySchdule(word roundKey){
	return (rotateLeft(roundKey, 15) ^ rotateLeft(roundKey, 32) ^ roundKey ^ 0x3);
}


uint16_t galoisMultiplication(uint16_t a, uint16_t b){// used it to find the lookup tables found in constants.h
	int p = 0;
    int hibitSet = 0;
    int i = 8;
    while(i--){
        if(b & 1){ 
        	p = p ^ a;
        }
        hibitSet = a & 0x80;
        a = a << 1;
        if(hibitSet == 0x80){
        	a = a ^ 0x1B;
        }
        b = b >> 1;
   	}
   	return p % 256;
}

column mixColumn(column c){
	column newColumn;
	newColumn.col[0] = galoisMul2[c.col[0]] ^ c.col[3] ^ c.col[2] ^ galoisMul3[c.col[1]];
	newColumn.col[1] = galoisMul2[c.col[1]] ^ c.col[0] ^ c.col[3] ^ galoisMul3[c.col[2]];
	newColumn.col[2] = galoisMul2[c.col[2]] ^ c.col[1] ^ c.col[0] ^ galoisMul3[c.col[3]];
	newColumn.col[3] = galoisMul2[c.col[3]] ^ c.col[2] ^ c.col[1] ^ galoisMul3[c.col[0]];
	return newColumn;
}




row sigma1(row r0){
	row newR0 = 0;
	newR0 |= (r0 & 0x8000) >>  0;  // 0
	newR0 |= (r0 & 0x4000) >>  6;  // 1
	newR0 |= (r0 & 0x2000) >>  8;  // 2
	newR0 |= (r0 & 0x1000) >> 10;  // 3
	newR0 |= (r0 & 0x0800) <<  3;  // 4
	newR0 |= (r0 & 0x0400) <<  1;  // 5
	newR0 |= (r0 & 0x0200) >>  5;  // 6
	newR0 |= (r0 & 0x0100) >>  7;  // 7
	newR0 |= (r0 & 0x0080) <<  6;  // 8
	newR0 |= (r0 & 0x0040) <<  4;  // 9
	newR0 |= (r0 & 0x0020) <<  2;  // A
	newR0 |= (r0 & 0x0010) >>  4;  // b
	newR0 |= (r0 & 0x0008) <<  9;  // C
	newR0 |= (r0 & 0x0004) <<  7;  // D
	newR0 |= (r0 & 0x0002) <<  5;  // E
	newR0 |= (r0 & 0x0001) <<  3;  // F

	return newR0;
}





row sigma2(row r1){
	row newR1 = 0;
	newR1 |= (r1 & 0x8000) >> 12;  // 0
	newR1 |= (r1 & 0x4000) >>  2;  // 1
	newR1 |= (r1 & 0x2000) >>  4;  // 2
	newR1 |= (r1 & 0x1000) >>  6;  // 3
	newR1 |= (r1 & 0x0800) >>  9;  // 4
	newR1 |= (r1 & 0x0400) <<  5;  // 5
	newR1 |= (r1 & 0x0200) >>  1;  // 6
	newR1 |= (r1 & 0x0100) >>  3;  // 7
	newR1 |= (r1 & 0x0080) >>  6;  // 8
	newR1 |= (r1 & 0x0040) <<  8;  // 9
	newR1 |= (r1 & 0x0020) <<  6;  // A
	newR1 |= (r1 & 0x0010) >>  0;  // B
	newR1 |= (r1 & 0x0008) >>  3;  // C
	newR1 |= (r1 & 0x0004) << 11;  // D
	newR1 |= (r1 & 0x0002) <<  9;  // E
	newR1 |= (r1 & 0x0001) <<  7;  // F

	return newR1;
}


row sigma3(row r2){
	row newR2 = 0;
	newR2 |= (r2 & 0x8000) >>  8;  // 0
	newR2 |= (r2 & 0x4000) >> 14;  // 1
	newR2 |= (r2 & 0x2000) >>  0;  // 2
	newR2 |= (r2 & 0x1000) >>  2;  // 3
	newR2 |= (r2 & 0x0800) >>  5;  // 4
	newR2 |= (r2 & 0x0400) >>  7;  // 5
	newR2 |= (r2 & 0x0200) <<  3;  // 6
	newR2 |= (r2 & 0x0100) <<  1;  // 7
	newR2 |= (r2 & 0x0080) >>  2;  // 8
	newR2 |= (r2 & 0x0040) >>  4;  // 9
	newR2 |= (r2 & 0x0020) << 10;  // A
	newR2 |= (r2 & 0x0010) <<  4;  // B
	newR2 |= (r2 & 0x0008) <<  1;  // C
	newR2 |= (r2 & 0x0004) >>  1;  // D
	newR2 |= (r2 & 0x0002) << 13;  // E
	newR2 |= (r2 & 0x0001) << 11;  // F

	return newR2;
}




row sigma4(row r3){
	row newR3 = 0;
	newR3 |= (r3 & 0x8000) >>  4;  // 0
	newR3 |= (r3 & 0x4000) >> 10;  // 1
	newR3 |= (r3 & 0x2000) >> 12;  // 2
	newR3 |= (r3 & 0x1000) <<  2;  // 3
	newR3 |= (r3 & 0x0800) >>  1;  // 4
	newR3 |= (r3 & 0x0400) >>  3;  // 5
	newR3 |= (r3 & 0x0200) >>  9;  // 6
	newR3 |= (r3 & 0x0100) <<  5;  // 7
	newR3 |= (r3 & 0x0080) <<  2;  // 8
	newR3 |= (r3 & 0x0040) >>  0;  // 9
	newR3 |= (r3 & 0x0020) >>  2;  // A
	newR3 |= (r3 & 0x0010) <<  8;  // B
	newR3 |= (r3 & 0x0008) <<  5;  // C
	newR3 |= (r3 & 0x0004) <<  3;  // D
	newR3 |= (r3 & 0x0002) <<  1;  // E
	newR3 |= (r3 & 0x0001) << 15;  // F

	return newR3;
}






word bitPermutationMixColumns(word w){
	row row0 = sigma1((w >> 48) & 0xFFFF);
	row row1 = sigma2((w >> 32) & 0xFFFF);
	row row2 = sigma3((w >> 16) & 0xFFFF);
	row row3 = sigma4((w >> 0 ) & 0xFFFF);


	//combined bit permutation with mixColumns so that I won't need to reconstruct the state and immidiatly break it up into rows again. 
	column col0, col1;

	col0.col[0] = ((row0&0xFF00)>>8);
	col0.col[1] = ((row1&0xFF00)>>8);
	col0.col[2] = ((row2&0xFF00)>>8);
	col0.col[3] = ((row3&0xFF00)>>8);
	



	col1.col[0] = (row0&0x00FF);
	col1.col[1] = (row1&0x00FF);
	col1.col[2] = (row2&0x00FF);
	col1.col[3] = (row3&0x00FF);
	


	col0 = mixColumn(col0);
	col1 = mixColumn(col1);


	word newWord = 0;

	newWord |= ((word)((col0.col[0] << 8) | col1.col[0]) << 48);
	newWord |= ((word)((col0.col[1] << 8) | col1.col[1]) << 32);
	newWord |= ((word)((col0.col[2] << 8) | col1.col[2]) << 16);
	newWord |= ((word)((col0.col[3] << 8) | col1.col[3]) <<  0);
	return newWord;

}


word applySbox(word w){
	word newW = 0;
	for(int j = 0; j < 64; j += 8){
		int byte = (w >> j) & 0xFF;
		newW |= ((word)sbox[byte]) << j;
	} 
	return newW;
}

word roundFunction(word w, word roundKey){
	
	w = w ^ roundKey;
	w = applySbox(w);
	w = bitPermutationMixColumns(w);
	
	return w;
}

word encrypt(word w, word masterKey, int rounds=7){
	word roundKey = masterKey;
	while(rounds--){
		w = roundFunction(w, roundKey);
		roundKey = keySchdule(roundKey);
	}
	return w;
}

uint64_t rdtsc(){
    unsigned int lo,hi;
    __asm__ __volatile__ ("rdtsc" : "=a" (lo), "=d" (hi));
    return ((uint64_t)hi << 32) | lo;
}



int main(){
	profileTime();
	return 0;
	word state = 0x0123456789ABCDEF;
	word firstRoundkey = 0x00000000FEDCBA98;
	cout << hex << uppercase << setfill('0') << setw(16) << state << endl;

	state = state ^ firstRoundkey;
	cout << hex << uppercase << setfill('0') << setw(16) << state << endl;

	state = applySbox(state);
	cout << hex << uppercase << setfill('0') << setw(16) << state << endl;

	state = bitPermutationMixColumns(state);
	cout << hex << uppercase << setfill('0') << setw(16) << state << endl;

	cout << hex << uppercase << setfill('0') << setw(16) << 0 << " " << hex << uppercase << setfill('0') << setw(16) << 0  << " " << hex << uppercase << setfill('0') << setw(16) << encrypt(0, 0) << endl;
	cout << hex << uppercase << setfill('0') << setw(16) << 0x42 << " " << hex << uppercase << setfill('0') << setw(16) << 0x1 << " " << hex << uppercase << setfill('0') << setw(16) << encrypt(0x42, 0x1) << endl;
	cout << hex << uppercase << setfill('0') << setw(16) << 0x0123456789ABCDEF << " " << hex << uppercase << setfill('0') << setw(16) << 0x00000000FEDCBA98 << " " << hex << uppercase << setfill('0') << setw(16) <<  encrypt(0x0123456789ABCDEF, 0x00000000FEDCBA98) << endl;
	return 0;
}




void profileTime(){
	word key = 0;
	uint64_t startClock = rdtsc();
	clock_t startTime = clock();
	clock_t startTime1 = clock();
	
	while(encrypt(0x0123456789ABCDEF, key) != 0xF0FE14D1C8C16C75){
		key++;
		if((key&0xFFFFFF)==0){
			uint64_t currentClock = rdtsc();
			clock_t currentTime = clock();
			double timer = ((double)(currentTime - startTime1))/CLOCKS_PER_SEC;
			double efficiancy1 = ((double)(currentClock-startClock))/0x1000000;
			double efficiancy2 = 0x1000000/timer;
			cout << "Key            : " << hex << uppercase << setfill('0') << setw(16) << key << endl;
			cout << "Efficiancy(e/s): " << efficiancy2 << " encryption/seconds" << endl;
			cout << "Efficiancy(c/e): " << efficiancy1 << " cycles/encryption" << endl;
			startClock = rdtsc();
			startTime1 = clock();
		}
	}
	cout << "Key found  : " << hex << uppercase << setfill('0') << setw(16) << key << endl;
	cout << "attack took: "	<< dec << (clock()-startTime)/CLOCKS_PER_SEC << " seconds" << endl;	
}

File: project/cipherImplementation.py
#!/usr/Bin/env python3

sbox = [0x63, 0x7C, 0x77, 0x7B, 0xF2, 0x6B, 0x6F, 0xC5, 0x30, 0x01, 0x67, 0x2B, 0xFE, 0xD7, 0xAB, 0x76,
		0xCA, 0x82, 0xC9, 0x7D, 0xFA, 0x59, 0x47, 0xF0, 0xAD, 0xD4, 0xA2, 0xAF, 0x9C, 0xA4, 0x72, 0xC0,
		0xB7, 0xFD, 0x93, 0x26, 0x36, 0x3F, 0xF7, 0xCC, 0x34, 0xA5, 0xE5, 0xF1, 0x71, 0xD8, 0x31, 0x15,
		0x04, 0xC7, 0x23, 0xC3, 0x18, 0x96, 0x05, 0x9A, 0x07, 0x12, 0x80, 0xE2, 0xEB, 0x27, 0xB2, 0x75,
		0x09, 0x83, 0x2C, 0x1A, 0x1B, 0x6E, 0x5A, 0xA0, 0x52, 0x3B, 0xD6, 0xB3, 0x29, 0xE3, 0x2F, 0x84,
		0x53, 0xD1, 0x00, 0xED, 0x20, 0xFC, 0xB1, 0x5B, 0x6A, 0xCB, 0xBE, 0x39, 0x4A, 0x4C, 0x58, 0xCF,
		0xD0, 0xEF, 0xAA, 0xFB, 0x43, 0x4D, 0x33, 0x85, 0x45, 0xF9, 0x02, 0x7F, 0x50, 0x3C, 0x9F, 0xA8,
		0x51, 0xA3, 0x40, 0x8F, 0x92, 0x9D, 0x38, 0xF5, 0xBC, 0xB6, 0xDA, 0x21, 0x10, 0xFF, 0xF3, 0xD2,
		0xCD, 0x0C, 0x13, 0xEC, 0x5F, 0x97, 0x44, 0x17, 0xC4, 0xA7, 0x7E, 0x3D, 0x64, 0x5D, 0x19, 0x73,
		0x60, 0x81, 0x4F, 0xDC, 0x22, 0x2A, 0x90, 0x88, 0x46, 0xEE, 0xB8, 0x14, 0xDE, 0x5E, 0x0B, 0xDB,
		0xE0, 0x32, 0x3A, 0x0A, 0x49, 0x06, 0x24, 0x5C, 0xC2, 0xD3, 0xAC, 0x62, 0x91, 0x95, 0xE4, 0x79,
		0xE7, 0xC8, 0x37, 0x6D, 0x8D, 0xD5, 0x4E, 0xA9, 0x6C, 0x56, 0xF4, 0xEA, 0x65, 0x7A, 0xAE, 0x08,
		0xBA, 0x78, 0x25, 0x2E, 0x1C, 0xA6, 0xB4, 0xC6, 0xE8, 0xDD, 0x74, 0x1F, 0x4B, 0xBD, 0x8B, 0x8A,
		0x70, 0x3E, 0xB5, 0x66, 0x48, 0x03, 0xF6, 0x0E, 0x61, 0x35, 0x57, 0xB9, 0x86, 0xC1, 0x1D, 0x9E,
		0xE1, 0xF8, 0x98, 0x11, 0x69, 0xD9, 0x8E, 0x94, 0x9B, 0x1E, 0x87, 0xE9, 0xCE, 0x55, 0x28, 0xDF,
		0x8C, 0xA1, 0x89, 0x0D, 0xBF, 0xE6, 0x42, 0x68, 0x41, 0x99, 0x2D, 0x0F, 0xB0, 0x54, 0xBB, 0x16]




def rotateLeft(word, n, wordSize=64):
    mask = 2**wordSize - 1
    return ((word << n) & mask) | ((word >> (wordSize - n) & mask))


def nextRoundKey(roundKey):
	return (rotateLeft(roundKey, 15) ^ rotateLeft(roundKey, 32) ^ roundKey ^ 0x3)

def getRows(word):
    row0 = (word >> 48) & 0xFFFF 
    row1 = (word >> 32) & 0xFFFF
    row2 = (word >> 16) & 0xFFFF
    row3 = (word >> 0) & 0xFFFF
    return row0, row1, row2, row3


def sigma(word):
	"""
		Implementing the sigma permutation on the 16 bit word.
	"""

	newWord = 0
	newWord |= (word & 0x8000) >> 0 # 0
	newWord |= (word & 0x4000) >> 6 # 1
	newWord |= (word & 0x2000) >> 8 # 2
	newWord |= (word & 0x1000) >> 10# 3
	newWord |= (word & 0x0800) << 3 # 4
	newWord |= (word & 0x0400) << 1 # 5
	newWord |= (word & 0x0200) >> 5 # 6
	newWord |= (word & 0x0100) >> 7 # 7
	newWord |= (word & 0x0080) << 6 # 8
	newWord |= (word & 0x0040) << 4 # 9
	newWord |= (word & 0x0020) << 2 # A
	newWord |= (word & 0x0010) >> 4 # b
	newWord |= (word & 0x0008) << 9 # C
	newWord |= (word & 0x0004) << 7 # D
	newWord |= (word & 0x0002) << 5 # E
	newWord |= (word & 0x0001) << 3 # F

	return newWord


def bitPermutation(word):
	"""
		Shift rows implementation
	"""
	row0, row1, row2, row3 = getRows(word)
    # Applying bit initial permutation
	row0 = sigma(row0)
	row1 = sigma(row1)
	row2 = sigma(row2)
	row3 = sigma(row3)

    # apply the shiftrows transformation(to make sure that no active bits will stay only in one column)
	row0 = row0 
	row1 = rotateLeft(row1, 4, 16)
	row2 = rotateLeft(row2, 8, 16) 
	row3 = rotateLeft(row3, 12, 16) 
    
    # reConstruCt the word
	newWord  = row0 << 48      # a |= b <==> a = a | b
	newWord |= row1 << 32
	newWord |= row2 << 16
	newWord |= row3 << 0
    
	return newWord


def galoisMult(a, b):
    """
    MultipliCation in the Galois field GF(2^8).
    """
    p = 0
    hibitSet = 0
    for i in range(8):
        if b & 1 == 1: 
        	p ^= a
        hibitSet = a & 0x80
        a <<= 1
        if hibitSet == 0x80:
        	a ^= 0x1B
        b >>= 1
    return p % 256

def mixColumn(column):
	"""
		The AES mix column for a single Column
	"""
	newCol = [0, 0, 0, 0]
	newCol[0] = galoisMult(column[0],2) ^ galoisMult(column[3],1) ^ galoisMult(column[2],1) ^ galoisMult(column[1],3)
	newCol[1] = galoisMult(column[1],2) ^ galoisMult(column[0],1) ^ galoisMult(column[3],1) ^ galoisMult(column[2],3)
	newCol[2] = galoisMult(column[2],2) ^ galoisMult(column[1],1) ^ galoisMult(column[0],1) ^ galoisMult(column[3],3)
	newCol[3] = galoisMult(column[3],2) ^ galoisMult(column[2],1) ^ galoisMult(column[1],1) ^ galoisMult(column[0],3)
	return newCol

def mixColumns(word):
	"""
		Implementation of the mix columns operation on "AES Mini"(which is the AES mixColumns but with only two columns instead of four)
	"""
	row0, row1, row2, row3 = getRows(word)
	column0 = []
	column1 = []

	column0.append((row0&0xFF00)>>8)
	column0.append((row1&0xFF00)>>8)
	column0.append((row2&0xFF00)>>8)
	column0.append((row3&0xFF00)>>8)
	
	

	column1.append(row0&0x00FF)
	column1.append(row1&0x00FF)	
	column1.append(row2&0x00FF)
	column1.append(row3&0x00FF)
	

	column0 = mixColumn(column0)
	column1 = mixColumn(column1)

	
	newWord = 0

	for i in range(4):
		newWord |= (((column0[i] << 8) | column1[i]) << ((4-i-1)*16))

	return newWord


def applySbox(word, sbox):
	"""
		apply the sbox to every byte 
	"""
	wordNew = 0
	for i in range(8): # 8 bytes
		j = i * 8
		byte = (word >> j) & 0xFF # retrieve the ith byte
		# insert the permuted byte in the CorreCt position
		wordNew |= sbox[byte] << j 
	return wordNew


def roundFunction(word, roundKey):
	"""
		The "AES Mini" round function
	"""
	
	word = word ^ roundKey
	word = applySbox(word, sbox)
	word = bitPermutation(word)
	word = mixColumns(word)
	

	return word


def encrypt(word, masterKey, rounds=7):
	roundKey = masterKey
	for i in range(rounds):
		# apply the roundfunction to word 
		word = roundFunction(word, roundKey)
		# go to the next round key
		roundKey = nextRoundKey(roundKey)
	return word




def create_test_vectors():

	state = 0x0123456789ABCDEF
	firstRoundkey = 0x00000000FEDCBA98
	print("%016X" % state)

	state = state ^ firstRoundkey
	print("%016X" % state)

	state = applySbox(state, sbox)
	print("%016X" % state)

	state = bitPermutation(state)
	# print("%016X" % state)

	state = mixColumns(state)
	print("%016X" % state)

	print("%016X" % 0, "%016X" % 0, "%016X" % encrypt(0, 0))
	print("%016X" % 0x42, "%016X" % 0x1, "%016X" % encrypt(0x42, 0x1))
	print("%016X" % 0x0123456789ABCDEF, "%016X" % 0x00000000FEDCBA98, "%016X" % encrypt(0x0123456789ABCDEF, 0x00000000FEDCBA98))




if __name__ == "__main__":

	import sys
	import hashlib
	import random

    
	if len(sys.argv) == 1:
		create_test_vectors()
		# print("Error occured")
		exit()

	key = int(sys.argv[1], 16)
	# We seed the random generator with a hash of the key to get the same messages for the same key
	random.seed(hashlib.sha256(sys.argv[1].encode()).digest())

	for i in range(16):
		word = random.getrandbits(64)
		cipher = encrypt(word, key, rounds=4)
		print("%016X %016X" % (word, cipher))
